tools/speedy_sync/src/main.rs

```diff
@@ -0,0 +1,272 @@
+struct Cli {
+    #[clap(subcommand)]
+    subcmd: CliSubcmd,
+}
+
+fn read_block_checkpoint(store: &Store, block_hash: &CryptoHash) -> BlockCheckpoint {
+    let block: Block = store
+        .get_ser(DBCol::Block, block_hash.as_ref())
+        .expect(format!("DB error Block {:?}", block_hash).as_str())
+        .expect(format!("Key missing Block {}", block_hash).as_str());
+
+    let info: BlockInfo = store
+        .get_ser(DBCol::BlockInfo, block_hash.as_ref())
+        .expect(format!("DB error BlockInfo {:?}", block_hash).as_str())
+        .expect(format!("Key missing BlockInfo {}", block_hash).as_str());
+
+    let merkle_tree: PartialMerkleTree = store
+        .get_ser(DBCol::BlockMerkleTree, block_hash.as_ref())
+        .expect(format!("DB error BlockMerkleTree {:?}", block_hash).as_str())
+        .expect(format!("Key missing BlockMerkleTree {}", block_hash).as_str());
+
+    BlockCheckpoint { header: block.header().clone(), info, merkle_tree }
+}
+
+fn write_block_checkpoint(store_update: &mut StoreUpdate, block_checkpoint: &BlockCheckpoint) {
+    let hash = block_checkpoint.header.hash();
+    store_update
+        .set_ser(DBCol::BlockHeader, hash.as_ref(), &block_checkpoint.header)
+        .expect("Failed writing a header");
+
+    store_update
+        .insert_ser(DBCol::BlockInfo, hash.as_ref(), &block_checkpoint.info)
+        .expect("Failed writing a block info");
+
+    store_update
+        .set_ser(DBCol::BlockMerkleTree, hash.as_ref(), &block_checkpoint.merkle_tree)
+        .expect("Failed writing merkle tree");
+    store_update
+        .set_ser(
+            DBCol::BlockHeight,
+            &index_to_bytes(block_checkpoint.header.height()),
+            block_checkpoint.header.hash(),
+        )
+        .unwrap();
+}
+
+fn write_epoch_checkpoint(store_update: &mut StoreUpdate, epoch_checkpoint: &EpochCheckpoint) {
+    store_update
+        .set_ser(DBCol::EpochInfo, epoch_checkpoint.id.as_ref(), &epoch_checkpoint.info)
+        .expect("Failed to write epoch info");
+}
+
+fn create_snapshot(create_cmd: CreateCmd) {
+    let path = Path::new(&create_cmd.home);
+    let store = StoreOpener::with_default_config().read_only(true).home(path).open();
+
+    // Get epoch information:
+    let mut epochs = store
+        .iter(DBCol::EpochInfo)
+        .filter_map(|(key, value)| {
+            if key.as_ref() == AGGREGATOR_KEY {
+                None
+            } else {
+                let info = EpochInfo::try_from_slice(value.as_ref()).unwrap();
+                let id = EpochId::try_from_slice(key.as_ref()).unwrap();
+                Some(EpochCheckpoint { id, info })
+            }
+        })
+        .collect::<Vec<EpochCheckpoint>>();
+
+    assert!(epochs.len() > 4, "Number of epochs must be greater than 4.");
+
+    epochs.sort_by(|a, b| a.info.epoch_height().partial_cmp(&b.info.epoch_height()).unwrap());
+    // Take last two epochs
+    let next_epoch = epochs.pop().unwrap();
```

Comment:
I think we should use epochs[-3] as `next_epoch`, because of the reason I mentioned that the chain must contains block headers up to two epochs old. That means, `block` must be at least two epochs away from the current head of the chain. Say the last epoch in the store is epoch T. That means, current chain head is higher than the last block of epoch T-2. So if we pick epochs[-3] which is epoch T-2 to be next_epoch, that means block_hash will be the last block of epoch T-4, and that's two epochs away from the current chain head.